planner/cascades/transformation_rules.go

```diff
@@ -1444,3 +1445,64 @@ func (r *TransformLimitToTableDual) OnTransform(old *memo.ExprIter) (newExprs []
 	tableDualExpr := memo.NewGroupExpr(tableDual)
 	return []*memo.GroupExpr{tableDualExpr}, true, true, nil
 }
+
+// PushLimitDownOuterJoin pushes Limit through Join.
+type PushLimitDownOuterJoin struct {
+	baseRule
+}
+
+// NewRulePushLimitDownOuterJoin creates a new Transformation PushLimitDownOuterJoin.
+// The pattern of this rule is `Limit -> Join`.
+func NewRulePushLimitDownOuterJoin() Transformation {
+	rule := &PushLimitDownOuterJoin{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandLimit,
+		memo.EngineTiDBOnly,
+		memo.NewPattern(memo.OperandJoin, memo.EngineTiDBOnly),
+	)
+	return rule
+}
+
+// Match implements Transformation interface.
+func (r *PushLimitDownOuterJoin) Match(expr *memo.ExprIter) bool {
+	if expr.GetExpr().HasAppliedRule(r) {
+		return false
+	}
+	join := expr.Children[0].GetExpr().ExprNode.(*plannercore.LogicalJoin)
+	return join.JoinType.IsOuterJoin()
+}
+
+// OnTransform implements Transformation interface.
+// This rule tries to pushes the Limit through outer Join.
+func (r *PushLimitDownOuterJoin) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	limit := old.GetExpr().ExprNode.(*plannercore.LogicalLimit)
+	join := old.Children[0].GetExpr().ExprNode.(*plannercore.LogicalJoin)
+	joinSchema := old.Children[0].Group.Prop.Schema
+	leftGroup := old.Children[0].GetExpr().Children[0]
+	rightGroup := old.Children[0].GetExpr().Children[1]
+
+	switch join.JoinType {
+	case plannercore.LeftOuterJoin, plannercore.LeftOuterSemiJoin, plannercore.AntiLeftOuterSemiJoin:
+		leftGroup = pushLimitDownOuterJoinToChild(limit, leftGroup)
+	case plannercore.RightOuterJoin:
+		rightGroup = pushLimitDownOuterJoinToChild(limit, rightGroup)
+	default:
+		return nil, false, false, nil
+	}
+
+	newJoinExpr := memo.NewGroupExpr(join)
+	newJoinExpr.SetChildren(leftGroup, rightGroup)
+	newLimitExpr := memo.NewGroupExpr(limit)
+	newLimitExpr.SetChildren(memo.NewGroupWithSchema(newJoinExpr, joinSchema))
+	newLimitExpr.AddAppliedRule(r)
+	return []*memo.GroupExpr{newLimitExpr}, true, false, nil
```

Comment:
maybe we can not erase the old plan. take this case as an example:

```sql
select * from s left join t on s.a = t.a limit 10;
```

if the row count of table `s` is smaller than 10, then the new plan `Table Scan(s) -> Limit(10)` is worse than `Table Scan(s)`.

Reply:
But considered that the `Limit` operator doesn't consume much resource during the execution phase and its execution time is really fast. I think it's OK to greedily erase the old plan:
* it's not that worse in the above case
* it's really worse when the row count of table `s` is so huge and there is not a `Limit(10)` after `Table Scan(s)`

Reply:
if `limit` be pushed down , `limit->gather->tableScan` will be generate, but not `tableScan->limit`.

Reply:
What you just said can be considered in `pushLimitDownGather`.
compare `limit.count` and `gather.DataSource.tableStats.RowCount`

Reply:
> if `limit` be pushed down , `limit->gather->tableScan` will be generate, but not `tableScan->limit`.

Correct. Here `Table Scan(s)` is only a logical speaking. In this rule, I think it should be `Any(on table s) -> Limit(10) -> OuterJoin(s, t) -> Limit(10)`

Reply:
> What you just said can be considered in `pushLimitDownGather`.
> compare `limit.count` and `gather.DataSource.tableStats.RowCount`

Yes, but without this rule to push the limit down through the join, there can not be a change that matches the pattern `xxGather -> Limit`. these rules are correlated.

Reply:
Maybe there can add a new attribute `maxRows` in plan like `LogicalPlan.maxRowsPerPartition`.
`gather.maxRows = gather.DataSource.tableStats.RowCount`;
`limit.maxRows = min(child.maxRows, limit.count)`;
`union.maxRows = sum(child.maxRows )`;

`stats.RowCount` is estimated value. `plan.maxRows` is the maximum value

Reply:
actually we have this mechanism in the execution engine: if the parent only required `maxRows`, the child is better to only return `maxRows`.